crates/sargon/src/signing/collector/signatures_collector_preprocessor.rs

```diff
@@ -0,0 +1,149 @@
+use crate::prelude::*;
+
+pub(crate) struct SignaturesCollectorPreprocessor {
+    transactions: IndexSet<TXToSign>,
+}
+
+pub(crate) fn sort_group_factors(
+    used_factor_sources: HashSet<FactorSource>,
+) -> IndexSet<FactorSourcesOfKind> {
+    let factors_of_kind: HashMap<FactorSourceKind, IndexSet<FactorSource>> =
+        used_factor_sources
+            .into_iter()
+            .into_grouping_map_by(|x| x.factor_source_kind())
+            .collect::<IndexSet<_>>();
+
+    let mut factors_of_kind = factors_of_kind
+        .into_iter()
+        .map(|(k, v)| (k, v.into_iter().sorted().collect::<IndexSet<_>>()))
+        .collect::<IndexMap<FactorSourceKind, IndexSet<FactorSource>>>();
+
+    factors_of_kind.sort_keys();
+
+    factors_of_kind
+        .into_iter()
+        .map(|(k, v)| {
+            FactorSourcesOfKind::new(k, v)
+                .expect("All factors should be of the same kind, since this is calling iter on a Map, using kind as key. Did you just move around lines of code?")
+        })
+        .collect::<IndexSet<_>>()
+}
+
+impl SignaturesCollectorPreprocessor {
+    pub(super) fn new(transactions: IndexSet<TXToSign>) -> Self {
+        Self { transactions }
+    }
+
+    pub(super) fn preprocess(
+        self,
+        profile_factor_sources: IndexSet<FactorSource>,
+    ) -> (Petitions, IndexSet<FactorSourcesOfKind>) {
+        let transactions = self.transactions;
+        let mut petitions_for_all_transactions =
+            IndexMap::<IntentHash, PetitionForTransaction>::new();
+
+        // We care for only the factor sources which are HD based
+        let mut all_factor_sources_in_profile =
```

Comment:
I think we should rather throw an Err if any Factor is not HD. Making this whole method  return `Result`.

Which wont affect anything since `SignaturesCollector` ctor already returns Result.